src/include/daos_api.h

```diff
@@ -220,11 +220,149 @@ daos_prop_alloc(uint32_t entries_nr);
 /**
  * Free the DAOS properties.
  *
- * \param[int]	prop	properties to be freed.
+ * \param[in]	prop	properties to be freed.
  */
 void
 daos_prop_free(daos_prop_t *prop);
 
+/**
+ * Allocate an DAOS Access Control List.
+ *
+ * \param[in]	aces		Array of pointers to ACEs to be put in the ACL.
+ * \param[in]	num_aces	Number of ACEs in array
+ *
+ * \return	allocated daos_acl pointer, NULL if failed
+ */
+struct daos_acl *
+daos_acl_alloc(struct daos_ace *aces[], uint16_t num_aces);
+
+/**
+ * Free a DAOS Access Control List.
+ *
+ * \param[in]	acl	ACL pointer to be freed
+ */
+void
+daos_acl_free(struct daos_acl *acl);
+
+/**
+ * Get the first Access Control Entry in the Access Control List, for iterating
+ * over the list.
+ *
+ * \param[in]	acl	ACL to traverse
+ *
+ * \return	Pointer to the first ACE in the ACL, or NULL if none exists
+ */
+struct daos_ace *
+daos_acl_get_first_ace(struct daos_acl *acl);
+
+/**
+ * Get the next Access Control Entry in the Access Control List, for iterating
+ * over the list.
+ *
+ * \param[in]	acl		ACL to traverse
+ * \param[in]	current_ace	Current ACE, to determine the next one
+ *
+ * \return	Pointer to the next ACE in the ACL, or NULL if at the end
+ */
+struct daos_ace *
+daos_acl_get_next_ace(struct daos_acl *acl, struct daos_ace *current_ace);
+
+/**
+ * Search the Access Control List for an Access Control Entry for a specific
+ * principal.
+ *
+ * \param[in]	acl		ACL to search
+ * \param[in]	type		Principal type to search for
+ * \param[in]	principal	Principal name, if type is USER or GROUP. NULL
+ *				otherwise.
+ *
+ * \return	Pointer to the matching ACE, or NULL if not found
+ */
+struct daos_ace *
+daos_acl_get_ace_for_principal(struct daos_acl *acl,
+		enum daos_acl_principal_type type, const char *principal);
+
+/**
+ * Insert an Access Control Entry in the appropriate location in the ACE
+ * list. The expected order is: Owner, Users, Assigned Group, Groups, Everyone.
+ *
+ * This requires us to reconstruct and reallocate the ACL structure. Both the
+ * old structure and the new one must be freed by the caller. The old data is
+ * left unchanged.
+ *
+ * If the new ACE is an update of an existing entry, it will replace the old
+ * entry.
+ *
+ * \param[in]	acl	Original ACL
+ * \param[in]	new_ace	ACE to be added
+ * \param[out]	new_acl	Reallocated copy of the ACL with new item inserted
+ *
+ * \return	DER_SUCCESS or error code
```

Comment:
Since this is DAOS API, forgive me for nit-picking. AFAICS, we don't actually use DER_SUCCESS, neither in APIs or in implementations; just 0 suffices. (ABT_SUCCESS is used because that's considered beyond our control.)

Reply:
correct and the "or error code" isn't super useful. We should list all possible errors to be returned.

Reply:
Got it. Fixed up the doxygen and added the error codes. Thanks for the clarification. I'd been using DER_SUCCESS since we define it.